src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,274 @@
+
+    // OpenSSL bindings do not expose custom extensions
+    // Parse the vcek using x509_parser
+    let vcek_der = &vcek.to_der()?;
+    let parsed_vcek = X509Certificate::from_der(vcek_der)?.1.tbs_certificate;
+
+    // verify vcek fields
+    // chip id
+    if get_oid::<64>(&parsed_vcek, HW_ID_OID)? != evidence.attestation_report.chip_id {
+        return Err(anyhow!("Chip ID mismatch"));
+    }
+
+    // tcb version
+    // these integer extensions are 3 bytes with the last byte as the data
+    if get_oid_int(&parsed_vcek, UCODE_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.microcode
+    {
+        return Err(anyhow!("Microcode verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, SNP_SPL_OID)? != evidence.attestation_report.reported_tcb.snp {
+        return Err(anyhow!("SNP verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, TEE_SPL_OID)? != evidence.attestation_report.reported_tcb.tee {
+        return Err(anyhow!("TEE verion mismatch"));
+    }
+
+    if get_oid_int(&parsed_vcek, LOADER_SPL_OID)?
+        != evidence.attestation_report.reported_tcb.boot_loader
+    {
+        return Err(anyhow!("Boot loader verion mismatch"));
+    }
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29f];
+
+    sig.verify(data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn load_milan_cert_chain() -> Result<(x509::X509, x509::X509)> {
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ask_ark.pem"))?;
+
+    // ask, ark
+    Ok((certs[0].clone(), certs[1].clone()))
+}
+
+fn verify_cert_chain(cert_chain: &[CertTableEntry]) -> Result<x509::X509> {
+    let (ask, ark) = load_milan_cert_chain()?;
+
+    let raw_vcek = cert_chain
+        .iter()
+        .find(|c| c.cert_type == SnpCertType::VCEK)
+        .ok_or_else(|| anyhow!("VCEK not found."))?;
+    let vcek = x509::X509::from_der(raw_vcek.data()).map_err(|_| anyhow!("Failed to load VCEK"))?;
+
+    // ARK -> ARK
+    ark.verify(&(ark.public_key().unwrap() as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ARK Signature"))
+        .unwrap();
+
+    // ARK -> ASK
+    ask.verify(&(ark.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid ASK Signature"))?;
+
+    // ASK -> VCEK
+    vcek.verify(&(ask.public_key()? as PKey<Public>))
+        .map_err(|_| anyhow!("Invalid VCEK Signature"))?;
+
+    Ok(vcek)
+}
+
+fn calculate_expected_report_data(nonce: &String, tee_pubkey: &TeePubKey) -> [u8; 64] {
+    let mut hasher = Sha384::new();
+
+    hasher.update(nonce.as_bytes());
+    hasher.update(&tee_pubkey.k_mod);
+    hasher.update(&tee_pubkey.k_exp);
+
+    let mut hash = hasher.finalize().to_vec();
+    hash.extend([0u8; 16]);
+
+    hash.as_slice().try_into().unwrap()
```

Comment:
nit: if you want to get rid of the result/unwrap you should be able to copy the slice content directly:

```suggestion
    let partial_hash = hasher.finalize();

    let mut hash = [0u8; 64];
    hash[..48].copy_from_slice(&partial_hash);
    hash
```

Reply:
Good suggestion. Thanks